pythoncode/FortiusAntBody.py

```diff
@@ -739,102 +732,63 @@ def Tacx2Dongle(self):
-            #       required power may change with speed, for same grade
-            #------------------------------------------------------------------
-            if TargetMode == mode_Grade:
-                r = usbTrainer.Grade2Resistance(TargetGrade, UserAndBikeWeight, SpeedKmh, Cadence)
-                r *= PowercurveFactor
-                TargetPower = usbTrainer.Resistance2Power(r, SpeedKmh)
+                if   TacxTrainer.Buttons == usbTrainer.EnterButton:     pass
+                elif TacxTrainer.Buttons == usbTrainer.DownButton:      TacxTrainer.SetPowercurveFactorDown()
+                elif TacxTrainer.Buttons == usbTrainer.UpButton:        TacxTrainer.SetPowercurveFactorUp()
+                elif TacxTrainer.Buttons == usbTrainer.CancelButton:    self.RunningSwitch = False
+                else:                                       pass
 
             #-------------------------------------------------------------------
-            # Prepare
+            # Do ANT work every 1/4 second
             #-------------------------------------------------------------------
             messages = []       # messages to be sent to ANT
             data = []           # responses received from ANT
-
-            #-------------------------------------------------------------------
-            # Send data to trainer (either power OR grade)
-            #-------------------------------------------------------------------
-            if False and clv.SimulateTrainer:   # I DO NOT REMEMBER NOW
-                pass                            # WHY THIS IS NOT DONE LIKE THIS
-
-            elif clv.Tacx_iVortex and VTX_VortexID:
-                if QuarterSecond:
-                    # Resistance = usbTrainer.Power2Resistance(TargetPower, SpeedKmh, Cadence)
-                    info = ant.msgPage16_TacxVortexSetPower (ant.channel_VTX_s, VTX_VortexID, TargetPower) # Resistance?
-                    messages.append ( ant.ComposeMessage (ant.msgID_BroadcastData, info) )
-
-            elif devTrainer:
-                usbTrainer.SendToTrainer(devTrainer, usbTrainer.modeResistance, \
-                    TargetMode, TargetPower, TargetGrade, PowercurveFactor, \
-                    UserAndBikeWeight, \
-                    PedalEcho, SpeedKmh, Cadence, Calibrate, clv.SimulateTrainer)
-
             if QuarterSecond:
                 LastANTtime = time.time()
-                #-------------------------------------------------------------------
+                #---------------------------------------------------------------
+                # Send power to iVortex
+                # Yes, we could have put this into the clsTacxTrainer
+                #      now all the ANT traffic is here.
+                #---------------------------------------------------------------
+                if clv.Tacx_iVortex and VTX_VortexID:
+                    info = ant.msgPage16_TacxVortexSetPower (ant.channel_VTX_s, VTX_VortexID, self.TargetPower)

```

Comment:
Will you refactor this into Trainer.SetPower() so that you can implement SetGrade() in terms of it? It seemed to refactor ok in that way.

Reply:
Conceptually, SetPower()/SetGrade() sets the data received from CTP into clsTacxTrainer.

Refresh() calls SendToTrainer() and should do the job and the code could be moved there.
See comment, I left it here. 

Thanks for pointing me to this code; self.TargetPower must be changed into TacxTrainer.TargetPower.

Reply:
And yes, I agree. The iVortex trainer class does not work without this external assistence.
But now the ANT-loop is kept in one place.

It's all about choices